src/backend/distributed/utils/causal_clock.c

```diff
@@ -0,0 +1,759 @@
+	TupleDesc tupleDescriptor = RelationGetDescr(pgDistCommitTransaction);
+	HeapTuple heapTuple = heap_form_tuple(tupleDescriptor, values, isNulls);
+
+	CatalogTupleInsert(pgDistCommitTransaction, heapTuple);
+
+	CommandCounterIncrement();
+
+	/* close relation and invalidate previous cache entry */
+	table_close(pgDistCommitTransaction, NoLock);
+}
+
+
+/*
+ * Initialize the shared memory clock value to the highest clock
+ * persisted. This will protect from any clock drifts.
+ */
+void
+InitClockAtBoot(void)
+{
+	if (creating_extension)
+	{
+		/* No catalog present yet */
+		return;
+	}
+
+	SpinLockAcquire(&logicalClockShmem->clockMutex);
+
+	/* Avoid repeated initialization */
+	if (logicalClockShmem->clockInitialized == true)
+	{
+		SpinLockRelease(&logicalClockShmem->clockMutex);
+		return;
+	}
+
+	/*
+	 * Set the flag before executing a distributed query
+	 * (else it might trigger this routine recursively.)
+	 */
+	logicalClockShmem->clockInitialized = true;
+
+	/* Start with the wall clock value */
+	uint64 epochValue = GetEpochTimeMs();
+	logicalClockShmem->clusterClockValue = START_NEW_CLOCK(epochValue);
+
+	SpinLockRelease(&logicalClockShmem->clockMutex);
+
+	/*
+	 * Version checking is a must for the below code as it accesses the
+	 * new catalog created in the new version.
+	 * Note: If we move this code to the top of the routine it will cause an
+	 * infinite loop with CitusHasBeenLoaded() calling this routine again.
+	 */
+	bool saveEnableVersionChecks = EnableVersionChecks;
+	EnableVersionChecks = true;
+
+	if (!CheckCitusVersion(NOTICE))
+	{
+		EnableVersionChecks = saveEnableVersionChecks;
+		return;
+	}
+
+	EnableVersionChecks = saveEnableVersionChecks;
+
+	/*
+	 * Select the highest clock value persisted in the catalog.
+	 *
+	 * SELECT i, MAX(j) FROM <tab>
+	 * WHERE i = (SELECT MAX(i) FROM <tab>)
+	 * GROUP BY i
+	 */
+	uint32 numCols = 2;
+	uint64 *results = ExecuteQueryAndReturnResults(
+		"SELECT clocklogical, clockcounter FROM "
+		"pg_catalog.pg_dist_commit_transaction ORDER BY 1 DESC, "
+		"2 DESC LIMIT 1;", numCols, SPI_OK_SELECT);
+
+	if (results != NULL)
+	{
+		uint64 logicalMaxClock = results[0];
+		uint32 counterMaxClock = results[1];
+
+		ereport(LOG,
+				(errmsg("Adjusted the clock with value persisted"
+						"logical(%lu) and counter(%u)",
+						logicalMaxClock, counterMaxClock)));
+
+		/*
+		 * Adjust the local clock according to the most recent
+		 * clock stamp value persisted in the catalog.
+		 */
+		AdjustLocalClockToRemote(logicalMaxClock, counterMaxClock);
+	}
+
+	/*
+	 * NULL results indicate no prior commit timestamps on this node, start
+	 * from the wall clock.
+	 */
+}
+
+
+/*
+ * ExecuteQueryAndReturnResults connects to SPI, executes the query and checks
+ * if the SPI returned the correct type. Returns an array of int64 results
+ * in the caller's memory context.
+ */
+static uint64 *
+ExecuteQueryAndReturnResults(char *query, int resultSize, int spiok_type)
```

Comment:
Would be nice to emphasize that this is only supposed to "return a sef of bigint columns for a single row", but not sure how.

Maybe, ExecuteQueryAndReturnResults -> ExecuteQueryAndReturnBigintCols ?

Reply:
Done.